landlab/utils/source_tracking_algorithm.py

```diff
@@ -0,0 +1,189 @@
+# %%
+# Import required libraries
+import numpy as np
+import copy
+from collections import Counter
+
+# %%
+def convert_arc_flow_directions_to_landlab_node_ids(grid, flow_dir_arc):
+    """
+    Convert Arc flow_directions file to be represented in node ids
+    gives receiver of each node starting at right and going clockwise.
+    
+    Parameters:
+    ----------
+    grid: RasterModelGrid
+        A grid.
+    flow_dir_arc: array_like
+        numpy.array of shape=[grid.number_of_nodes]; flow directions derived
+        from ESRII ArcGIS.
+    
+    Returns:
+    -------
+    receiver_nodes: array_like
+        A numpy.array of shape=[grid.number_of_nodes]. Please note that this
+        array gives the receiver nodes only for the core nodes. For non-core
+        nodes, a zero is used.  
+    """
+    r_arc_raw = np.log2(flow_dir_arc)
+    r_arc_raw = r_arc_raw.astype('int')
+    neigh_ = grid.neighbors_at_node
+    diag_ = grid.diagonals_at_node
+    neigh_ = np.fliplr(neigh_)
+    diag_ = np.fliplr(diag_)
+    a_n = np.hsplit(neigh_, 4)
+    a_d = np.hsplit(diag_, 4)
+    neighbors = np.hstack((a_n[-1], a_d[0], a_n[0], a_d[1], a_n[1], a_d[2],
+                           a_n[2], a_d[3]))
+    # Now neighbors has node ids of neighboring nodes in cw order starting at
+    # right, hence the order of neighbors = [r, br, b, bl, l, tl, t, tr]
+    receiver_nodes = np.zeros(grid.number_of_nodes, dtype=int)
+    receiver_nodes[grid.core_nodes] = np.choose(r_arc_raw[grid.core_nodes],
+                                   np.transpose(neighbors[grid.core_nodes]))
+    return (receiver_nodes)
+
+
+# %%
+# Source Routing Algorithm
+# Note 1: This algorithm works on core nodes only because core nodes
+# have neighbors that are real values and not -1s.
+# Note 2: Nodes in the following comments in this section refer to core nodes.
+def track_source(grid, hsd_ids, flow_directions=None):
+    """
+    This algorithm traverses the grid, and records all upstream nodes for each
+    node, given the flow directions. Alternatively, flow directions can be
+    attached to the grid as a field.
+    
+    Parameters:
+    ----------
+    grid: RasterModelGrid
+        A grid.
+    hsd_ids: array_like
+        numpy.array of shape=[grid.number_of_nodes]; array that maps the
+        nodes of the grid to, possibly coarser, Hydrologic Source Domain (HSD)
+        grid ids.
+    flow_directions: array_like, optional.
+        numpy.array of shape=[grid.number_of_nodes]; array of receivers.
+        Alternatively, this data can be provided as a nodal field
+        'flow__receiver_node' on the grid.
+    
+    Returns:
+    -------
+    (hsd_upstr, flow_accum): (dictionary, array_like)
+        'hsd_upstr' maps each Model Domain (MD) grid node to corresponding
+        contributing upstream HSD ids. 'flow_accum' is an array of the number
+        of upstream contributing nodes.
+    """   
+    if flow_directions==None:
+        r = grid.at_node['flow__receiver_node']
```

Comment:
We will see a field error if no flow directions are provided. Hence, leaving it for now. 
@kbarnhart: I will work on adding a call to D8 flow router in a future commit.